core/arch/arm/sm/psci-helper.S

```diff
@@ -0,0 +1,156 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <kernel/unwind.h>
+#include <asm.S>
+#include <arm.h>
+#include <arm32_macros.S>
+
+/*
+ * From ARM DDI 0406C.c Page B2-1286
+ * Example code for cache maintenance operations
+ *
+ * flush dcache all to PoC
+ */
+FUNC psci_armv7_flush_dcache_all , :
+UNWIND(	.fnstart)
+	stmfd	sp!, {r4-r11, lr}
+	dmb
+
+	/*
+	 * Get cache coherency level
+	 * If level is 0, do not need to clean.
+	 */
+	read_clidr r0
+	isb
+	ands	r3, r0, #0x7000000
+	mov	r3, r3, lsr #23
+	beq	finished
+
+	/* For each cache level */
+	mov	r10, #0
+loop1:
+	/*
+	 * r0: clidr value
+	 * r2: the Ctype[x] shift bits
+	 * r10: the current cache level
+	 */
+	add	r2, r10, r10, lsr #1
+	mov	r1, r0, lsr r2
+	and	r1, r1, #7
+	cmp	r1, #2
+	blt	skip
+
+	write_csselr r10
+	isb
+	read_ccsidr r1
+
+	/* Get line size */
+	and	r2, r1, #7
+	add	r2, r2, #4
+
+	/* Get number of ways */
+	movw	r4, #0x3ff
+	ands	r4, r4, r1, lsr #3
+	clz	r5, r4
+	/* R9 working copy of the max way size */
+	mov	r9, r4
+
+loop2:
+	/* Get max index */
+	movw	r7, #0x7fff
+	ands	r7, r7, r1, lsr #13
+loop3:
+	orr	r11, r10, r9, lsl r5
+	orr	r11, r11, r7, lsl r2
+	write_dccisw r11
+	subs	r7, r7, #1
+	bge	loop3
+	subs	r9, r9, #1
+	bge	loop2
+
+skip:
+	add	r10, r10, #2
+	cmp	r3, r10
+	bgt	loop1
+finished:
+	mov	r10, #0
+	write_csselr r10
+	dsb	st
+	isb
+	ldmfd	sp!, {r4-r11, lr}
+	bx	lr
+UNWIND(	.fnend)
+END_FUNC psci_armv7_flush_dcache_all
+
+FUNC psci_disable_smp, :
+UNWIND(	.fnstart)
+	read_actlr r0
+	bic	r0, r0, #(1 << 6)
+	write_actlr r0
+	isb
+	bx	lr
+UNWIND(	.fnend)
+END_FUNC psci_disable_smp
+
+FUNC psci_enable_smp, :
+UNWIND(	.fnstart)
+	read_actlr r0
+	orr	r0, r0, #(1 << 6)
+	write_actlr r0
+	isb
+	bx	lr
+UNWIND(	.fnend)
+END_FUNC psci_enable_smp
+
+FUNC psci_armv7_cpu_off, :
+UNWIND(	.fnstart)
+	push	{lr}
+
+	bl	psci_armv7_flush_dcache_all
+
+	clrex
```

Comment:
Actually I have no good idea on this. From my experience of PSCI U-Boot interface, I think this is needed. I am not sure this is correct or not:  This core may execute LDREX/STREX before, so when offline, need to execute CLREX to clear-exclusive.